### PR TITLE
Move priority queue to new common crate

### PR DESCRIPTION
**Pull request by agourlay**
This PR belongs to the ongoing work to support sparse vectors.

The [experimental implementation](https://github.com/qdrant/sparse-vectors-experiments) is based on the `FixedLengthPriorityQueue` in this repo. 

In order to keep components decoupled, it will be added in a different crate.

To avoid having a dependency on `segment`, the queue is pushed in a new common crate which will host more things in the future.

### CODE REVIEW

File: lib/segment/src/spaces/tools.rs [hunk 1]
```diff
@@ -1,115 +1,6 @@
...
-
-    fn into_iter(self) -> Self::IntoIter {
-        IntoIter {
-            it: self.heap.into_sorted_vec().into_iter(),
-        }
-    }
-}
+use common::fixed_length_priority_queue::FixedLengthPriorityQueue;
 
 pub fn peek_top_smallest_iterable<I, E: Ord>(elements: I, top: usize) -> Vec<E>
```
Comment (agourlay) [inline lib/segment/src/spaces/tools.rs:5]: I did not move those utils function to keep the diff simple.

It might make sense to move those later.

**Comment by IvanPleshkov** [comment]
I propose to create a separate repo for common stuff. I find separate crate easier to support and I can move something to this crate for the quantization library to avoid unnecessary conversions

**Comment by agourlay** [comment]
@IvanPleshkov Can you please elaborate why you think this should be an external crate? 
Is this only about the quantization crate?

**Comment by generall** [comment]
> I propose to create a separate repo for common stuff

Don't agree. Quantization is more of the exception, cause it required a lot of experiments. At this moment, I think nobody except for you can do changes in quantization repo. 
Common stuff should stay within qdrant repo

**Comment by IvanPleshkov** [comment]
> @IvanPleshkov Can you please elaborate why you think this should be an external crate? Is this only about the quantization crate?

In this pipeline I cannot downgrade in cargo.toml sparse vector library version (there may be some reasons to do that, for testing as example). I find that downgrade of sparse vector library may conflict with head qdrant version. If common crate is a separate repo, I can downgrade both libraries in qdrant. Correct me if I'm wrong

**Comment by IvanPleshkov** [comment]
I expect that sparse vectors will be implemented as a separate repo (like repo with sparse vector experiments). If not, all my topics was wrong

**Comment by agourlay** [comment]
@IvanPleshkov  The goal is to have sparse vectors as a separate crate **within** the main repo.

The common crate is a way to decouple our internal crates for faster compile times and cleaner interfaces.

Is this making things clearer?

**Comment by IvanPleshkov** [comment]
Thanks for the explanation, I thought that sparse vectors will be implemented separately like WAL. Now it's clear and I don't have any concerns

**Review by generall — APPROVED** [review]

**Pull request merged**